### ISSUE TITLE
[Feature Request] High Availability for vpc nat gateway

### ISSUE THREAD

**Issue by zhangzujian** [kubeovn/kube-ovn#4415]
### Description

Implement high availability for vpc nat gateway, including iptables snat and eip.

We can:

1. Run a DaemonSet instead of StatefulSet;
2. Use lease to do a leader election;
3. The leader pod is responsible for announcing eip via ARP/NDP or BGP;
4. Use lsp options `requested-chassis` to change port binding.

**Comment by bobz965** [comment]
1. Run a DaemonSet instead of StatefulSet;

Do you mean to run DaemonSet pod on each node?

**Comment by SkalaNetworks** [comment]
Would running a deployment not be lighter? In any case, we need to have the content of the netfilter synced to avoid distruption when a node/nat gw pod crashes. If not possible, it will still be "HA", but there's gonna be TCP connection breakages on node/pod failures.

My use case is a NAT GW that can be "migrated" between nodes, the same way you'd migrate a VM. This is useful to drain a node for maintenance for example. Also right now, if a node crashes with a natgw on it, the nat gw will never come back because of statefulset mechanics in Kubernetes.

**Comment by zhangzujian** [comment]
> In any case, we need to have the content of the netfilter synced to avoid distruption when a node/nat gw pod crashes. If not possible, it will still be "HA", but there's gonna be TCP connection breakages on node/pod failures.

Currently we have no ability to sync conntrack entries between pods, so after a node/pod failure existing, connections will be broken and reconnection is needed.

**Comment by zhangzujian** [comment]
> Could we use a tool like conntrackd and only enable it on "handovers" when we want to switch traffic from one nat-gw to another? This mode could work great for migrations. Another mode could be implemented with permanent synchronization for very-HA usages where random crashes of nodes/pods should not cause any breakage.

`conntrackd` seems ok.

**Comment by SkalaNetworks** [comment]
How do we ensure compatibility with "old" gateways? Rollout a new API version?

**Comment by SkalaNetworks** [comment]
Hey, starting the thread again because I think this is a really important subject for Kube-OVN.

Right now, VPC NAT GW have the following issues:
- They're Statefulsets, which means updating them forces the recreation of the Pod with a downtime roughly equal to 30s
- They're Statefulsets, which means if a node on which they are crashes, the pod will NOT be brought back on another node until the node is back or a human intervenes. This is terrible because there's not just a loss of connectivity for a couple seconds, the entire VPC/Subnet gets disconnected from the Internet until something is done.
- They're Statefulsets, so you can't have multiple running with the same state, through conntrackd for example
- They can't work active/passive, active/active

My proposal for a design solution:
- conntrackd running in UDP or multicast to synchronize the states
- opt-in active/active, active/passive by default, the kube-ovn-controller handles handover when it sees a gateway has failed (how does it know though? if it must wait for the node on which it is to be marked as NotReady, it might take a long time)
- active/active obviously needs the GW to run in BGP mode to have equalized traffic using whatever hashing the router supports through ECMP
- active/active could be implemented using leases to detect node failures quickly
- everything is a deployment, can set anti-affinity to get them to different places
- rollout updates are supported, when a pod is signaled to be killed, it voluntarily relinquishes its active status if it is active, and only terminates once it has confirmed another pod has taken active status

**Comment by bobz965** [comment]
@SkalaNetworks Do you mean to use a deployment instead of sts for better active/active nat gw pods design ？

**Comment by SkalaNetworks** [comment]
Yes, statefulsets never come back up when a node is down by design. Deployments will.

**Comment by bobz965** [comment]
> Yes, statefulsets never come back up when a node is down by design. Deployments will.

LGTM！

**Comment by SkalaNetworks** [comment]
@zhangzujian Hi, I'll probably start working on the feature this week, I need HA NAT gateways in production, as running everything off one Pod is not ideal, especially since node crashes need manual interventions from operators to reschedule the statefulsets.

I'd like some guidance considering I'm in no way an OVN/OVS pro, so I'll ask a few questions here hoping you can help me :)

1. We need to spread the traffic of type "egress", coming from the VPC and going out (probably to the Internet) by passing through the NAT gateway

You seem to have dealt with that in the VPC Egress Gateway feature, using (I think) ECMP+BFD, ECMP for consistent hashed spreading of the traffic and BFD to detect when a gateway fails and redirect traffic elsewhere using ECMP. From what I gather of the code, it seems pretty simple to do. I just need to define a BFD IP and have a daemon on each gateway.

Do you think this should be how the NAT gateway detects failures? Your initial post included Leader/Followers for detection. The drawback from this is slower detection of failures and the setup is Active/Passive instead of Active/Active (which would be a better setup).

If I implement ECMP+BFD, I still have a question. During rehashing, traffic will switch from a failed node to a working one. In your implementation of the VPC Egress Gateway, doesn't it lead to TCP RST due to changing the egress IP? This shouldn't be the case on the NAT gateway, considering every gateway will have the same EIPs and SNAT/DNAT rules, right?

2. This brings me to my second question, which is syncing conntrack states. I don't think the VPC Egress Gateway does it, so there's probably gonna be cutoffs on failures and rehash. I'm kinda fine with that on the NAT gateway, but I'd like to explore a way to make it as resilient as possible. In that case, my initial proposal of using conntrackd for active/active conntrack syncing still stands. We would use multicast and IGMP snooping (I think it's something that can be enabled on Subnets within Kube-OVN) to share the conntrack within instances.  Do you think it's a good idea? Or has OVN features I don't know about which would help making a stateful HA firewall easier?


3. Concerning Ingress traffic, back into the NAT gateway, the best way to handle it is to use the BGP feature I implemented a few month back. ECMP would be handled by the remote speaker, spreading traffic on all the instances. There would also be a need for BFD here. How does the remote speaker know when a NAT gateway is down without waiting for the BGP timer to expire (could take several seconds)? I guess we would expose the BFD server on the "external subnet"?


If you've got a design proposal, I'm all hears. I'll start doing a proof of concept based on your code from the Egress Gateway.

**Comment by SkalaNetworks** [comment]
I'll handle this whenever I get the time, I'm working on a lot of different stuff at the same time and I couldn't get to it properly. I already have a concept of a PR and the refactor got merged.

**Comment by zbb88888** [comment]
@SkalaNetworks hi, is there any doc about this feature?

**Comment by zbb88888** [comment]
1. egress： 

vpc LRP（enable BFD） --> nat-gw-pod(internal nic + openBfdd) with lrp static ecmp route: a better way ???


2. ingress 

phsical switch  (enable BGP with  ECMP + BFD ) --> nat-gw-pod(external nic + openBfdd): a better way ???

**Comment by SkalaNetworks** [comment]
I'm not sure I understand your message, are you proposing an alternative to what was discussed here or are you making a call for better ideas?

BFD/ECMP was pretty established from using the same methods as the egress gateway. The logical gateway would then forward to the NAT gateways.

For the ingress, BGP/ECMP and BFD for fast failovers was also discussed but no definite answer on the engine for BFD (openbfdd could be a solution).

There still needs to be something to synchronize conntracks, otherwise if a node fails, rehashing will forward traffic to another NAT gateway where it will be TCP RST or dropped. For that I proposed conntrackd.

There's no documentation, only ideas here in this issue. I started a bit of refactoring of the NAT GW a few month back to make it easier to implement the HA feature, but didn't go much further than that. I probably have a branch somewhere with the transformation from an STS to a deployment though.

**Comment by zbb88888** [comment]
@SkalaNetworks ， Hi, it is not an alternative. I just post some details about what i want to add in the future, which all based on the info we have talked before.

**Comment by SkalaNetworks** [comment]
I am actively working on this feature, emulating the same behaviour that was implemented for the EgressGateway, with BFD, policy routes and the such. The difference will obviously be in how I synchronize the states of the NAT gateways.

**Comment by SkalaNetworks** [comment]
First phase was merged, next phase is working on a stable means of sharing the conntrack between instances
